src/resource.rs

```diff
@@ -139,6 +147,72 @@ impl Debug for LowerFunc {
     }
 }
 
+/// Sealed trait for shared type parametrised functionality between [`TypeDef`] and [`OpDef`].
+pub trait TypeParametrisedInternal: sealed::SealedDef {
+    /// The concrete object built by binding type arguments to parameters
+    type Concrete: CustomConcrete;
+    /// The resource-unique name.
+    fn name(&self) -> &SmolStr;
+    /// Type parameters.
+    fn params(&self) -> &[TypeParam];
+    /// The parent resource. if any.
+    fn resource(&self) -> Option<&ResourceId>;
+    /// Check provided type arguments are valid against parameters.
+    fn check_args(&self, args: &[TypeArg]) -> Result<(), SignatureError> {
+        for (a, p) in args.iter().zip(self.params().iter()) {
+            check_type_arg(a, p).map_err(SignatureError::TypeArgMismatch)?;
+        }
+        Ok(())
+    }
+
+    /// Instantiate a concrete instance by providing type arguments.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the provided arguments are not
+    /// valid instances of the TypeDef parameters.
+    fn to_custom(&self, args: impl Into<Vec<TypeArg>>) -> Result<Self::Concrete, SignatureError>;
```

Comment:
Let's rename this - how about `instantiate`? Or `to_concrete` if you must (we're calling instantiations concrete, which I'm happy with, but we could be consistent)